exchanges/kraken/kraken_websocket.go

```diff
@@ -286,303 +289,249 @@ func (k *Kraken) wsProcessSystemStatus(respRaw []byte) error {
+	Open                       [2]types.Number `json:"o"`
+}
 
-		f, err := strconv.ParseFloat(s, 64)
-		if err != nil {
-			return fmt.Errorf("received invalid ticker data: %w", err)
-		}
-		data[key] = f
+// wsProcessTickers converts ticker data and sends it to the datahandler
+func (k *Kraken) wsProcessTickers(dataRaw json.RawMessage, pair currency.Pair) error {
+	var t wsTicker
+	if err := json.Unmarshal(dataRaw, &t); err != nil {
+		return fmt.Errorf("error unmarshalling ticker data: %w", err)
 	}
 
 	k.Websocket.DataHandler <- &ticker.Price{
 		ExchangeName: k.Name,
-		Ask:          data["a"],
-		Bid:          data["b"],
-		Close:        data["c"],
-		Volume:       data["v"],
-		Low:          data["l"],
-		High:         data["h"],
-		Open:         data["o"],
+		Ask:          t.Ask[0].Float64(),
+		Bid:          t.Bid[0].Float64(),
+		Close:        t.Last[0].Float64(),
+		Volume:       t.Volume[0].Float64(),
+		Low:          t.Low[0].Float64(),
+		High:         t.High[0].Float64(),
+		Open:         t.Open[0].Float64(),
 		AssetType:    asset.Spot,
 		Pair:         pair,
 	}
 	return nil
 }
 
+type wsSpread struct {
+	Bid       types.Number
+	Ask       types.Number
+	Time      types.Time
+	BidVolume types.Number
+	AskVolume types.Number
+}
+
+func (w *wsSpread) UnmarshalJSON(data []byte) error {
+	return json.Unmarshal(data, &[5]any{&w.Bid, &w.Ask, &w.Time, &w.BidVolume, &w.AskVolume})
+}
+
 // wsProcessSpread converts spread/orderbook data and sends it to the datahandler
-func (k *Kraken) wsProcessSpread(response []any, pair currency.Pair) error {
-	data, ok := response[1].([]any)
-	if !ok {
-		return errors.New("received invalid spread data")
-	}
-	if len(data) < 5 {
-		return errors.New("unexpected wsProcessSpread data length")
-	}
-	bestBid, ok := data[0].(string)
-	if !ok {
-		return common.GetTypeAssertError("string", data[0], "bestBid")
-	}
-	bestAsk, ok := data[1].(string)
-	if !ok {
-		return common.GetTypeAssertError("string", data[1], "bestAsk")
-	}
-	timeData, ok := data[2].(string)
-	if !ok {
-		return common.GetTypeAssertError("string", data[2], "timeData")
-	}
-	timestamp, err := strconv.ParseFloat(timeData, 64)
-	if err != nil {
-		return err
-	}
-	bidVolume, ok := data[3].(string)
-	if !ok {
-		return common.GetTypeAssertError("string", data[3], "bidVolume")
+func (k *Kraken) wsProcessSpread(rawData json.RawMessage, pair currency.Pair) error {
+	var data wsSpread
+	if err := json.Unmarshal(rawData, &data); err != nil {
+		return fmt.Errorf("error unmarshalling spread data: %w", err)
 	}
-	askVolume, ok := data[4].(string)
-	if !ok {
-		return common.GetTypeAssertError("string", data[4], "askVolume")
-	}
-
 	if k.Verbose {
-		log.Debugf(log.ExchangeSys,
-			"%v Spread data for '%v' received. Best bid: '%v' Best ask: '%v' Time: '%v', Bid volume '%v', Ask volume '%v'",
+		log.Debugf(log.ExchangeSys, "%s Spread data for '%v' received. Best bid: '%v' Best ask: '%v' Time: '%v', Bid volume '%v', Ask volume '%v'",
```

Comment:
🤷‍♂️ Not definite, but can we use `%q` throughout this?
```suggestion
		log.Debugf(log.ExchangeSys, "%s Spread data for '%v' received. Best bid: '%v' Best ask: '%v' Time: '%v', Bid volume '%v', Ask volume '%v'",
```

Reply:
%q only works with strings or anything which implements a String() method but not for floats or anything else. These vars support the String method, so can switch to that since this is just a debug log